### ISSUE TITLE
[RFC][WIP] Hierarchical ContentTypes

### ISSUE THREAD

**Issue by graham73may** [bolt/bolt#5989]
It would be great to address the hierarchical content situation. Myself and @mrenigma are working on a solution for an existing project that follows what we’ve outlined below. Once we’ve finished our extended Bolt version it should be fairly easy to create as a PR for core. 

# Backend

## Parent selection drop down
Currently I'm using a standard select field of the same content type. The selected ID is saved in the database. I eventually want to get to a point where the options are displayed as follows:

Drop down to list root pages, with their descendants shown directly underneath. Depth shown by dashes. 

```
Grandparent Page
- Parent Page
- - Child Page
About
- History
- Staff
- - Board
- - Trustees
- Contact
Research
- Title
- Title
- Title
```

## Slug field displaying the correct lineage
The slug field will need to update on:
- load/save 
- parent select field change

### On load:
Using `POST_HYDRATE`, the contenttype is checked for `hierarchical: true`. The parent field is then checked to find the piece of contents lineage. 

Once the lineage has been determined, the slugs are used to construct a `route_prefix` which is added via a `POST_HYDRATE` function to the `contentType` object.

### On select change:
Not looked into this yet, Ajax.

## Multi-level nested pages
This is defined by the parents drop down, if you select a child page as the parent then you will be creating a grandchild page and so on...

## Allow same slugs to be used under different parents
Not looked into this yet. Also need to ensure same slugs are not used under the same parent. 

## Listing template
* Show sub-pages underneath the parent page. 
    * POA: add a new block to `_sub/listing.twig` to output sub pages under the root parent page https://github.com/bolt/bolt/blob/release/3.1/app/view/twig/_sub/_listing.twig
* Paging will ignore children pages (Sub pages will not count towards the paging limit)

## Don't allow a circular lineage
Not looked into this yet. 

# Frontend

## Routing
Add a `hierarchical: true/false` setting to the contenttypes.yml so the controller knows whether to check for parents

Add support for slashes `/` in URLs, current this works using the following:

```
pagelink:
    path: /{slug}
    defaults:
        _controller: soapbox.controller.pages:record
        contenttypeslug: page
    contenttype: pages
    requirements:
        slug: [a-z0-9-_/]+
```
This needs to be set up to extend the basic slug requirements internally to match `/` **only** when `hierarchical` is set to true instead.

### Routing functionality
The routing functionality breaks the URL down. Let’s use `/root-page/parent-page/child-page/` as an example. 

Firstly it will `getContent()` where `slug` === `root-page` AND `parent` field === `’’ || NULL`. 

Now it has the ID for `root-page` it can look for `slug` === `parent-page` AND `parent` === `{root-page ID}`

If any `getContent()` check on the URL fails, the page doesn’t exist, throw a 404. 

## Related link functionality
Check how this affects `Content:link()`

**Comment by graham73may** [comment]
Dynamically setting the `route_prefix` relies on this PR being merged: 
https://github.com/bolt/bolt/pull/5988

**Comment by SahAssar** [comment]
Just flagging that there is also a lot of previous discussion about this in #1295

**Comment by graham73may** [comment]
@mrenigma Just a few random thoughts about forking this into core:

- Any thoughts on making a new field type called `parentselect`?
     - The hierarchy code for the options is already being output, but at the moment the JS is tidying it inside the standard `select` field, if we had a separate field type we could output it however we wanted and it would have just the one purpose. 
    - Can we automatically add the `parentselect` field to the contenttype if `hierarchical:true`?
- When this is part of core, can we set the routing config based on `hierarchical:true`? I remember this not being possible as part of the extended implementation.
- There are currently bugs with the listing template in the backend - if you multiple select child items and press the action buttons at the bottom, nothing seems to be corrupted, but the listing is flattened and the order changes.

**Comment by ToBe998** [comment]
This is one of the most requested features here. I'd really like this getting implemented as it would dramatically help us generating proper urls for SEO purposes.

BUT: It isnt usable for any website that has more than one contenttype for it's content. And that means basically all websites i know of. So, how would you use that approach for websites that have more than one contenttype?
For example: We have articles and portals (routing pages) and products and videos and much more...

**Comment by graham73may** [comment]
@tobe998 multiple hierarchical contenttypes has worked fine on our extended uses so far. The URL hits the new routing functionality and that knows how to handle the requested item based on its URL, parents, slug, etc. 

This WIP is on hold at the moment whilst the new Storage class is being implemented. New PRs aren't being accepted on Legacy classes.

Also very keen to get hierarchical content into Bolt!

**Comment by ToBe998** [comment]
So you can just have multiple hirarchical routes like the above sample and all will try to find a slug of their contenttype. The first one that finds a match wins? 
That would be perfect. I highly doubt that two contenttypes have objects with same slug ;)

**Comment by graham73may** [comment]
The way I like to setup my sites (and maybe it does this out of the box, I can't remember) is so that pages don't have the contenttype in the URL. E.g.:

`/about-us/our-history/`

And other contenttypes do have the contenttype in the URL

`/publications/root-publication/chapter-name/`

**Comment by ToBe998** [comment]
Yea, that will work. And im certainly fine with this, but SEO isnt most of the time ;) 
So it's not possible to have more than one route without contenttype here?

**Comment by graham73may** [comment]
Might be missing something here, what's wrong with having /events/ at the start of the URL? Usually I would have /events/ as a listing/archive type page and then the events are "under" that.

**Comment by ToBe998** [comment]
This is slightly offtopic, so only the very short version: According to our seo guys, a url needs to follow these points as close as possible:
- as short as possible (with good slugs fine in this case)
- only human readable and content related parts (some contenttypes might make sense, but many dont. for example /block/my-favourite-block would be a bad contenttype)
- folders should mimic the content hirarchy with most important category/channel as first element. (This is actually the most difficult with bolt and an important thing for the seo guys as it makes analytics over certain content areas possible. Example /technology/batteries/a-new-litium-battery-invented)

**Comment by graham73may** [comment]
@ToBe998 My understanding is that you can setup `routing.yml` to do the URLs however you want just like normal. 

Block's don't have to be `/block/block-name`, you can change `routing.yml` so that they become `/chunks/block-name`

Either way, keep an eye on things and we can discuss when we have something to play with :-)

**Comment by graham73may** [comment]
This is not stale, the WIP is nearly ready to be PR'd

https://github.com/mrenigma/bolt/tree/feature/hierarchical-content

**Comment by GawainLynch** [comment]
Zero stress mate, it is a GitHub bot making the call :wink:

**Comment by graham73may** [comment]
@GawainLynch Just making sure it doesn't get auto closed :D

**Comment by GawainLynch** [comment]
@graham73may — yup, figured … just making sure I engage so we don't look like :hankey:s 

:rofl:

**Comment by rossriley** [comment]
Not stale, next on the list after #7291

**Comment by mrl22** [comment]
Sorry to bring this back up, I am not great with github issue tracking. Has this issue been resolved and implemented? It is a real deal-breaker for us as we have some large sites that need pages to be nested. One project that we are working on will have the same name as other pages within the site, for example /leisure/events and /business/events .

**Comment by bobdenotter** [comment]
It's not for 3, but it might be implemented for v4. 

If it's truly a dealbreaker, there are two ways you can help make it happen:

 - Help building this feature, by allocating manpower to help
 - Sponsor us financially: https://twitter.com/BoltCMS/status/1182565391173902341

**Comment by mrl22** [comment]
Sponsored :) I guess the CMS is not ready for us yet, keep up the good work.

**Comment by bobdenotter** [comment]
@mrl22 Truly appreciated! Let me know under which name & url you'd like to be mentioned in the README and admin screen. 

Also, thinking about your question.. Something we can do right now in v3, is to use this: https://github.com/xiaohutai/bolt-hierarchicalroutes

You can see the results on https://notaris.nl or https://transplantatiestichting.nl . If you'd like to know more, feel free to hop on Slack for more information.

**Comment by mrl22** [comment]
@bobdenotter Richard Leishman www.webfwd.co.uk is fine.

I did try bolt-hierarchicalroutes, but I am trying to create it user fiendly for the client also. Issue I have with bolt-hierarchicalroutes is that the slug shows incorrect on editing the page, the client will need to be taught to create the menu item and guess the full slug. Almost all of our projects are WordPress/ACF but we are increasingly getting requests for a non-wordpress cms due to security attacks. We love the ease of creating content-blocks and using other field-types.

I would support by helping to develop the feature for Bolt, but I am not the best with Classes/Namespaces.

**Comment by bobdenotter** [comment]
Ah, yes.. The visibility / usability of that in the current Bolt backend is not optimal, because there it's still presented as a "flat" bunch of records. 

> I would support by helping to develop the feature for Bolt, but I am not the best with Classes/Namespaces.

No pressure, but it's not all classes and namespaces when it comes to that. Even hanging out in Slack, helping the odd novice with a beginner question is part of that, imho. And everything in between. 😇

**Comment by graham73may** [comment]
We are using this PR for a number of production sites and it does have the
URL in a client friendly way in the backend. It has meant we are stuck on
an older core version of Bolt though so probably wouldn't recommend it.
